proto/private/rules/proto_toolchain.bzl

```diff
@@ -0,0 +1,57 @@
+# Copyright 2021 The Bazel Authors. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""A toolchain for Protobuf rules."""
+
+load("//proto/private:native.bzl", proto_common = "native_proto_common")
+
+def _proto_toolchain_impl(ctx):
+    return [
+        platform_common.ToolchainInfo(
+            protoc = ctx.attr._protoc[DefaultInfo].files_to_run,
+            protoc_options = getattr(ctx.fragments.proto, "protoc_options", []),
+        ),
+    ]
+
+def _protoc():
+    if "protoc" in getattr(proto_common, "available_configuration_fields", []):
+        return configuration_field(
+            fragment = "proto",
+            name = "protoc",
+        )
+    return "@com_google_protobuf//:protoc"
+
+proto_toolchain = rule(
+    implementation = _proto_toolchain_impl,
+    attrs = {
+        "_protoc": attr.label(
```

Comment:
Should this be a public attribute?

Reply:
It should, but `LateBoundDefaults` can't be used with public attributes :/